rlbottraining/common_exercises/kickoff_exercise.py

```diff
@@ -0,0 +1,122 @@
+
+from rlbot.matchcomms.common_uses.reply import send_and_wait_for_replies
+from rlbot.matchcomms.common_uses.set_attributes_message import make_set_attributes_message
+from rlbot.matchconfig.match_config import MatchConfig, PlayerConfig, Team
+from rlbot.utils.game_state_util import BallState, CarState, GameState, Physics, Rotator, Vector3
+
+from rlbottraining.common_graders.kickoff_grader import KickoffGrader
+from rlbottraining.grading.grader import Grader
+from rlbottraining.match_configs import make_empty_match_config
+from rlbottraining.paths import BotConfigs
+from rlbottraining.rng import SeededRandomNumberGenerator
+from rlbottraining.training_exercise import TrainingExercise, Playlist
+
+
+@dataclass
+class SpawnLocation:
+    pos: Vector3
+    rot: Rotator
+
+
+class Spawns():
+    """Default Kickoffs Spawns (BLUE)"""
+    CORNER_R = SpawnLocation(Vector3(-2048, -2560, 18), Rotator(0, 0.25*pi, 0))
+    CORNER_L = SpawnLocation(Vector3(2048, -2560, 18), Rotator(0, 0.75*pi, 0))
+    BACK_R = SpawnLocation(Vector3(-256, -3840, 18), Rotator(0, 0.5*pi, 0))
+    BACK_L = SpawnLocation(Vector3(256.0, -3840, 18), Rotator(0, 0.5*pi, 0))
+    STRAIGHT = SpawnLocation(Vector3(0, -4608, 18), Rotator(0, 0.5*pi, 0))
+
+
+@dataclass
+class KickoffExercise(TrainingExercise):
+    grader: Grader = field(default_factory=KickoffGrader)
+    blue_spawns: List[SpawnLocation] = field(default_factory=list)
+    orange_spawns: List[SpawnLocation] = field(default_factory=list)
+
+    def __post_init__(self):
+        """Flip the orange spawns around to get the correct location and combine the two spawns."""
+        orange_spawns = [
+            SpawnLocation(Vector3(-spawn.pos.x, -spawn.pos.y, spawn.pos.z),
+            Rotator(spawn.rot.pitch, spawn.rot.yaw+pi, spawn.rot.roll))
+            for spawn in self.orange_spawns]
+        self.spawns = self.blue_spawns + orange_spawns
+
+    def on_briefing(self):
+        """Sends a match comm to let your bot know a new kickoff is starting."""
+
+        """ Uncomment this first part if you want to use match comms.
+
+        _ = send_and_wait_for_replies(self.get_matchcomms(), [
+            make_set_attributes_message(0, {'kickoff': True}),
+        ])
+        """
+
+        """ Put this in your bot to recieve the match comms:
+
+        try:
+            msg = self.matchcomms.incoming_broadcast.get_nowait()
+        except Empty:
+            return
+        if handle_set_attributes_message(msg, self, allowed_keys=['kickoff']):
+            reply_to(self.matchcomms, msg)
+        else:
+            self.logger.debug(f'Unhandled message: {msg}')
+        """
+        pass
+
+    def make_game_state(self, rng: SeededRandomNumberGenerator) -> GameState:
+
+        num_players = self.match_config.num_players
+        assert num_players == len(self.spawns), 'Number of players does not match the number of spawns.'
+
+        car_states = {}
+        for index in range(num_players):

```

Comment:
The constraint about red/orange matching the match config isn't well enforced currently.
```suggestion
        for i, player in enumerate(self.match_config.player_configs):
            assert (player.team == BLUE) == (i < len(self.blue_spawns)), "blue/orange players in matchconfig doesn't match the blue/orange spawns. expected {} blue player_configs, followed by {} orange ones.".format(len(self.blue_spawns), len(self.orange_spawns))
```